### ISSUE TITLE
(failing doc) keep Cython-generated .c files in the repository?

### ISSUE THREAD

**Issue by dnerini** [pySTEPS/pysteps#99]
There is a problem with readthedocs failing to build the pysteps documentation:
https://readthedocs.org/projects/pysteps/builds/9398855/

The problem seems to have appeared with commit d760a639e7fb89754b8145fe4a802efdc0f44c89 ("Add compilation and interface for the Proesmans module").

Any idea what could be the cause? Seems to be related to the compilation of the cython code for the new Proesmans method.

**Comment by aperezhortal** [comment]
I think that this is caused by a missing optional dependency: Cython. In the setup.py, when Cython is not found, it will compile the extensions using the c sources:
```
_proesmans_lib_extension = Extension(str("pysteps.motion._proesmans"),
                                         sources=[str("pysteps/motion/_proesmans.c")],
                                         **_proesmans_extension_arguments)
```
The corresponding C source is not available in the repository. 

This opens an interesting discussion: should we add Cython as dependency and  remove the C sources. This will create the C sources during the installation process? 

The C sources generated by Cython constains metadata with the user's locals path and configurations (which you may not want to share). 

For example, Pandas use many Cython extensions, but it seems that the C generated files are not included. 
Here there is an interesting discussion on the topic: https://github.com/pyFFTW/pyFFTW/issues/31

**Comment by dnerini** [comment]
Thanks, @aperezhortal, the explanation makes a lot of sense!

Indeed, it opens an interesting discussion. Personally, I don't see major problems in having Cython as dependency and removing the C sources.

As a third option, we could also imaging keeping Cython as an optional dependency, removing C sources and simply raising a MissingOptionalDependency exception whenever a method requiring Cython is called (currently motion.vet and motion.proesmans only).

**Comment by aperezhortal** [comment]
> As a third option, we could also imaging keeping Cython as an optional dependency, removing C sources and simply raising a MissingOptionalDependency exception whenever a method requiring Cython is called (currently motion.vet and motion.proesmans only).

This will produce an error during the installation, since the modules need to be compiled prior the installation of the package. Once the package is installed, cython is no longer needed.

For example, the Pandas package in conda-forge includes Cython as a dependency in host section. https://github.com/conda-forge/pandas-feedstock/blob/master/recipe/meta.yaml

Since the package managers (pip or conda) handles the dependencies very well, adding an extra dependency shouldn't be an issue.

**Comment by dnerini** [comment]
> This will produce an error during the installation, since the modules need to be compiled prior the installation of the package. Once the package is installed, cython is no longer needed.

But in principle we could simply skip the compilation if cython is not installed.
Anyway, I agree that having cython as mandatory dependency probably represents the best solution.

**Comment by pulkkins** [comment]
I pushed the missing _proesmans.c to the repository. To me it looks awkward to have Cython-generated .c files in the repository (especially when they contain local user's local paths). I would prefer keeping Cython as an optional dependency because none of the C modules is essential for pysteps.

**Comment by dnerini** [comment]
Thanks @pulkkins, commit 63db885efc65a74ba50125da40114f7f143ecb17 fixed the problem with the failing documentation.

Regarding now the auto-generated C source code and the Cython dependency, in order to remove the former, we need to adopt the latter, as already discussed above.

That is to say, we have two main options:

1. keep C code in the repository, leave Cython as optional dependency (current status)
2. remove C code, adopt Cython as mandatory dependency (proposed by @aperezhortal )

there might be a third way where we remove the C code and keep Cython as optional dependency, but we don't know yet whether this is actually an option.

so, what do we do? Any more ideas/thoughts?

**Comment by kmuehlbauer** [comment]
> there might be a third way where we remove the C code and keep Cython as optional dependency, but we don't know yet whether this is actually an option.

@dnerini Actually the third option would be to adopt [PEP518](https://www.python.org/dev/peps/pep-0518/) and add Cython as a build requirement. Then the .c-files can be removed from the repo and Cython can be removed as direct (optional) dependency.

This would work for all `pip`-based installs. For those who clone the repo and want to use `python setup.py install` a `requirements_devel.txt` can be added if necessary.

**Comment by aperezhortal** [comment]
@kmuehlbauer, PEP518 is a good option. I was also experimenting with a new setup.py that delays the cythonization until the dependencies are installed by setup.py (install_requires or setup_requires keywords). The later fix will work for in `pip` and `python setup.py install`. But this something that PEP518 tries to avoid in the first place.  

> For those who clone the repo and want to use python setup.py install a requirements_devel.txt can be added if necessary.

I think that this is a good way to go. If the setup.py requirements are not installed (numpy and cython), we can raise an exception with a message indicating that the requirements needs to be installed.

In this branch I started the refactoring process of the build system where we can try this new ideas: 
https://github.com/pySTEPS/pysteps/tree/refactor_build_system

**Comment by kmuehlbauer** [comment]
@aperezhortal PEP518 is currently adopted by more and more packages. The nice thing for users is that their python environment will not be cluttered with the build-time dependencies.

**Comment by aperezhortal** [comment]
I agree. I was thinking about a solution that implements PEP518 and also checks (but do not install) the setup requirements if the user wants to install the package using `python setup.py install`. 

But, maybe we can recommend installing the package from sources using pip (instead calling setup.py)  (even for development): https://packaging.python.org/tutorials/installing-packages/#installing-from-a-local-src-tree

We can experiment in the [refactor_build_system branch](https://github.com/pySTEPS/pysteps/tree/refactor_build_system) a little bit and see how this goes.

**Comment by kmuehlbauer** [comment]
@aperezhortal Not sure if this is the best place for my comments regarding build system. In the [refactor_build_system_branch](https://github.com/pySTEPS/pysteps/tree/refactor_build_system) another travis run could be added which checks the install via pip.

**Comment by aperezhortal** [comment]
@kmuehlbauer, that's a good idea. We can test the installation from a minimal environment using pip.  I'll implement it and see how it goes. Thanks!

**Comment by kmuehlbauer** [comment]
@aperezhortal If you create a PR from that branch to master, discussion could go along there and we would have the ability to make comments within the changed code.

**Comment by kmuehlbauer** [comment]
Just got this on gitter: https://mobile.twitter.com/pganssle/status/1152694289920679936

Tldr; Do not use `python setup install`

**Comment by dnerini** [comment]
PR #102 closes this issue, thanks everyone for the nice and constructive discussion!
